doc/sphinx/language/core/assumptions.rst

```diff
@@ -118,6 +124,13 @@ Function application
    | @term1
 
 :n:`@term1__fun @term1` denotes applying the function :n:`@term1__fun` to :token:`term1`.
+The type of :n:`@term1__fun` must be a product, (`forall ...`), which is the type of
+a term that is a function.  The function may be known, such as `Nat.add`, or unknown,
+such as any hypothesis that doesn't have a
+:term:`local definition <context-local definition>`.
```

Comment:
I think your alternative is considerably worse.  It's not helpful if you don't know all the concepts.  I give the concrete details (e.g. what does the type look like, what does it mean, how do I create a valid function application) while yours just gets into a bit of theory (well-typedness) and doesn't get concrete.  AFAIK the doc doesn't explain when and why expressions have to be well-typed.  I don't know those rules. I'd expect that everything should be well typed.  I this should be explained elsewhere and with somewhat more explanation.

Typing is one of the worst-explained thing in the doc (unless you're already an expert on it).

Reply:
> I'd expect that everything should be well typed.

That's correct. The only thing that nuances this is that sometimes, checks for well-typedness are delayed (e.g., when using some unsafe tactics, checks for well-typedness are delayed until `Qed`).

FWIW, the doc already defines the term "well-typed": https://coq.inria.fr/refman/language/core/basic.html#term-well-typed

Reply:
> Typing is one of the worst-explained things in the doc (unless, perhaps, you're already an expert on it).

Personally, I find that the section https://coq.inria.fr/refman/language/core/basic.html#term-type (in which "well-typed" is defined) is pretty good, but of course, I'm already well-informed on these things, so I can't be the best judge.

Reply:
However, the doc doesn't have a good description (understandable by non-experts) of how to compute the type of a term.  Just listing the typing rules is nowhere near sufficient.  Nor do I understand when type checking is done or not done beyond at `Qed` time.

Perhaps improving the description of typing would be a good next task on the documentation.  (Explaining typeclasses and coercions may be good candidates, too--I really have no idea what they're useful for.)  But don't need to pick the next target today.

Reply:
> I'd expect that everything should be well typed.

I had not re-read the definition of well typed before writing this.

Reply:
In any case, regarding the changes at stake here, if we cannot find some common grounds, let's not change this paragraph. I do not like your proposal (too convoluted) and you do not like mine (too abstract).

Reply:
I guess we could drop it, but perhaps we can cover these points without too much discussion:

- As Jason pointed out, we don't define "function".  Is there a fundamental difference between functions whose type is "forall ... " (e.g. `plus_O_n`) versus "A -> B" (e.g `Nat.add`) ?  IIUC `fix` and `cofix` specified items can only have the latter type.  If so, we should distinguish these two cases.

- IIUC evaluation only applies to things in the `Nat.add` category, and reduction is the mechanism for doing so.  IIUC there is no evaluation of `plus_O_n`.  Is that right?

- Why do I get different types for these 2 cases that seem like they should give the same type?

```
Print plus_O_n.
(* plus_O_n = fun n : nat => eq_refl : forall n : nat, 0 + n = n *)
Ltac t1 := let t := type of plus_O_n in idtac t.
Ltac t2 := let t := type of
     (fun n : nat => eq_refl : forall n : nat, 0 + n = n) in idtac t.
Goal True.
t1.   (* (forall n : nat, 0 + n = n) *)
t2.
(* In environment
n : nat
The term "eq_refl" has type "?x = ?x"
while it is expected to have type
 "forall n : nat, 0 + n = n". *)
```

Reply:
>As Jason pointed out, we don't define "function".  Is there a fundamental difference between functions whose type is "forall ... " (e.g. `plus_O_n`) versus "A -> B" (e.g `Nat.add`) ?

No, the latter is just a subcase of the former.

>IIUC evaluation only applies to things in the `Nat.add` category, and reduction is the mechanism for doing so.  IIUC there is no evaluation of `plus_O_n`.  Is that right?

No, evaluation applies to any term and sometimes it is stuck. Evaluation is just the operation of running *any* reduction (see https://coq.inria.fr/refman/proofs/writing-proofs/equality.html#computing-in-a-term-eval-and-eval). It applies to any kind of functions applied to arguments using the beta-reduction rule, and to other sorts of terms using other reduction rules. The reason why `plus_O_n x` does not reduce is not because it has a dependent product type, but because `plus_O_n` is an opaque term (thus, Coq cannot unfold its body, much like if it were an axiom without a body). If `plus_O_n` had been `Defined` instead of `Qed`, then it would reduce when applied to arguments. So, the fact that something reduces or not is not related to any consideration that would be specifically related to functions.

>Why do I get different types for these 2 cases that seem like they should give the same type?

You don't get different types. You get the same type in the result of `Print` and `t1` here. The reason you get an error in the case of `t2` is because of missing parentheses. You cannot just copy the whole thing left of `=` in the output of `Print` and expect this to be a valid term because it is actually two parts. I.e., if I decompose the output of `Print` further, it is:

```
name = term : type
```

So the term here is just `fun n : nat => eq_refl`.

You can fix your definition of `t2` by removing the type:

```coq
Ltac t2 := let t := type of
     (fun n : nat => eq_refl) in idtac t.
```

Or by adding parentheses (to be allowed to keep the type, which is interpreted as a cast then):

```coq
Ltac t2 := let t := type of
     ((fun n : nat => eq_refl) : forall n : nat, 0 + n = n) in idtac t.
```

Reply:
The second fix for `t2` works, but your first fix for `t2` gets this error:

`Cannot infer the implicit parameter A of eq_refl whose type is "Type" in environment: n : nat`

Separately, why do `fix` and `{ struct }` appear in this definition?  The description of the `term_fix` nonterminal doesn't explain what these actually do or why they're useful.  (I'd like to do another pass on that chapter sometime.)

```
Nat.add = 
fix add (n m : nat) {struct n} : nat :=
  match n with
  | 0 => m
  | S p => S (add p m)
  end
     : nat -> nat -> nat
```

Reply:
`Nat.add` is a recursive function. It is defined using `Fixpoint`, which is just a convenience command for defining recursive functions (the same definitions could be obtained using `Definition` + `fix` and this is what is going on internally).